Feature/pyqtchart integration

bradyzp opened the pull request:
This branch contains the PyQtGraph integrations, some of the work here is deprecated due to our decision to move away from using PyQtGraph graphic flowcharts for the transformation workflow.

ToDo: Clean-up deprecated code after merging in the new transformation branch. Specifically features in the dgp/lib/transform package which will be superseded by the new transform workflows.

Features Added in this Branch:
- New Flight-Line selection plot based on PyQtGraph - this provides a much more responsive UI experience when viewing raw data to select flight lines.
- Flight Workspaces (dgp/gui/workspaces): A modular approach to handling the sub-tab widgets for each flight - allowing fast switching between multiple open flights and in future the ability to view & mutate data for each flight in its own tab. Workspaces can be dynamically loaded, and each is contained within its own python module.
- Added an excepthook method in the program __main__ module, and in the test runner config which captures and prints stack-traces for uncaught Qt exceptions, which otherwise crash with an integer error value.

bradyzp committed:
Cleanup and rework of Channel Selection Feature

CLN: Separate documentation requirements from main requirements.txt the
idea being to include only the core requirements to execute the main
program.
CLN: Moved ProjectTreeView class out of main.py into its own views.py
ENH: Changed working of channel selection for the main plot. The
ChannelList Tree View is now acessible via a dialog box above the plot.
This greatly reduces the complexity in the code as callbacks are no
longer required to tie back into the main window.
ENH: Add default channels to plot. The plot tab can now optionally plot
a few default channels (e.g. Gravity, Long/Cross) to the main plot area
if data is available when the tab is created.
CLN: Refactored some names in widgets.py to better reflect the classes'
purposes. E.g. TabWorkspace -> FlightWorkspace.

FIX: Added def to catch and print stacktrace when GUI crashes due to
unhandled python exception.

WIP: Plotter work. Fixes to Widgets.py.

Added central difference, eotvos, and test for eotvos.

Added gradient derivative for testing

Delete eotvos module and test

Added basic graph functionality for the backend and some tests.

bradyzp committed:
ENH: Added functionality to Pyqtgraph for transform node

Implemented basic DateFormatter for Pyqtgraph tick lines.
Added TransformPlot class to handle creation of PyQtgraph Plot Widgets
for transform interface - sets plots with DateAxis and links the x-axis
scales.
Change Node Library initialization - add module parameter that
determines whether to present Nodes for use by the user, or to hide them
as is the case for the Display/Plot nodes - which require special
instantiation.
Updated FIRLowpassFilter to include the Series name in its output (for
use with plot legend)

ENH: Further refinements to Pyqtgraph date formatter.

Improvements to Pyqtgraph formatter for transform plot interface.
Working on common Plot API for MPL/PG plots.

Improvements to Plotting backend, refactoring of GUI Tab Code.

Refactored dgp/gui/widgets.py and extracted individual Sub-Tab class
definitions into their own class file. This should ease the future
addition of different tabs, each as their own self contained classes
that can then be assembled within a workspace.

ENH: Added second flight-line selection plot using pyqtgraph.

Added improved Flight Line selection plot based on PyQtGraph library,
for better interactive performance.
Added line-selection toggle mode - user can click on/off to enter/exit
line selection/editing mode.

cbertinato committed:
ENH: Added functionality to align and crop gravity and trajectory frames (#64)

Add Override classes for line selection click handling.

Add LinearFlightRegion subclass of PyQtGraphs LinearRegionItem which is
used to replicate the old flight-line selection behavior allowing users
to click and drag/resize rectangular areas on the plot.
Integrated this into plotters.py to create new LFR's on click when in
line-selection mode.
TODO: Figure out how/where the best way to place labels on the regions
is.

CLN: Remove deprecated transform logic from Tab.

Preparing to merge into develop to continue development with new graph
code - removing dependence on PyQtGraph for the flowchart DAG.
Old graph logic and transform functions still remain, will merge changes
over after this branch is merged into develop.

Refactoring/Cleanup in gui/plotting interfaces.

Cleaned up some parts of the PyQtGraph backend and added documentation
to various methods.
Marked some old MPL interfaces as obsolete for later consideration
(deletion?)

Fix test failing due to dependency refactoring.

Update travis config to reflect new build_uic.py location.
Also moved build_uic.py utility script out of build directory, as this
directory is used by setuptools as a build location.

cbertinato commented on dgp/gui/dialogs.py:
```diff
@@ -416,9 +414,11 @@ def path(self, value):
             self.line_path.setText('None')
             return
 
+        print("Raw path value: ", value)
```
For debug, or still need this?

cbertinato commented on dgp/gui/main.py:
```diff
@@ -234,37 +215,39 @@ def _launch_tab(self, index: QtCore.QModelIndex=None, flight=None) -> None:
+        tab = self._flight_tabs.widget(index)  # type: FlightTab
+        if tab.subtab_widget():
+            print("Active flight subtab has a widget")
+        # self._context_tree.setModel(tab.context_model)
```
Can we remove these commented lines? Also below at 249 and 250?

cbertinato commented on dgp/gui/plotting/mplutils.py:
```diff
@@ -380,6 +390,95 @@ def reset_view(self, x_margin=None, y_margin=None):
+                group.set_edge(edge, 'red', select=False)
+                break
+        else:
+            # group.set_edge('', 'black', select=False)
```
Still need this commented code?

bradyzp replied:
Nope, cleaned that out. In fact I'm not sure about the future of the mplutils classes, they might be useful for plotting transformed data, but like below with plotters.py will have to consider whether to keep them around.

cbertinato commented on dgp/gui/plotting/plotters.py:
```diff
@@ -1,45 +1,326 @@
+
+
+"""
+TODO: Many of the classes here are not used, in favor of the PyQtGraph line selection interface.
```
I think it's ok to leave this around until we work out the remaining issues with the new plotting interface.

cbertinato commented on dgp/gui/views.py:
```diff
@@ -0,0 +1,105 @@
+        confirm.setWindowTitle("Confirm Delete")
+        res = confirm.exec_()
+        if res:
+            print("Emitting item_removed signal")
```
Should this print statements be moved into the logs?

bradyzp replied:
Removed those as I was just using them to debug something there, turned the exception handler into a logging statement.

cbertinato commented on dgp/gui/workspace.py:
```diff
@@ -0,0 +1,131 @@
+        self._line_proc_tab = LineProcessTab("Line Processing", flight)
+        self._workspace.addTab(self._line_proc_tab, "Line Processing")
+
+        # self._map_tab = WorkspaceWidget("Map")
```
Can this be removed?

cbertinato commented on examples/pyqtgraph_line_selection_plot.py:
```diff
@@ -0,0 +1,96 @@
+
+        self._plot = LineSelectPlot(flight=self._flight, rows=3)
+        self._plot.line_changed.connect(lambda upd: print(upd))
+        # self.plot.figure.canvas.mpl_connect('pick_event', lambda x: print(
```
Can these be removed?

cbertinato commented on examples/pyqtgraph_line_selection_plot.py:
```diff
@@ -0,0 +1,96 @@
+        xrng = x1 - x0
+        tenpct = xrng * .1
+
+        # lri = LinearRegionItem()
```
And here?

cbertinato reviewed (changes requested):
Nice work, Zach! I haven't done any QA testing yet, but if anything arises when I do, then I'll just make an issue.

bradyzp committed:
CLN: Cleaned up old comments/debug statements.

Removed debug print statements and commented debug code.
Changed data file delete context action to warn (log) that the function
is not yet implemented (instead of raising a NotImplementedError)
Removed unused code paths from workspaces.py from old tree-context
switching between flight tabs - flights now have their own embedded
'model' to choose data channels.
